### PR TITLE
feat(next-api): basic authentication setup

### PR DESCRIPTION
**Pull request by raisedadead**
This recreates #49329 with conflicts resolved

Rel: #49327
Closes: #49472

### CODE REVIEW

File: api/index.ts [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import { config } from 'dotenv';
 config({ path: '../.env' });
 import fastifyAuth0 from 'fastify-auth0-verify';
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
```
Comment (ojeytonwilliams) [inline api/index.ts:5]:
```suggestion
// @ts-expect-error declaration files will be provided
```

File: api/routes/test.ts [hunk 2]
```diff
@@ -21,6 +21,9 @@ export async function testRoutes(fastify: FastifyInstance) {
           _res,
           done
         ) {
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          // @ts-ignore
```
Comment (ojeytonwilliams) [inline api/routes/test.ts:25]:
```suggestion
          // @ts-expect-error the types will be added
```
ts-expect-error is self-correcting. Once the error goes away, _it_ will create an error, forcing you to remove it.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: api/routes/test.ts (hunk 2 above)
Comment (Nirajn2311) [reply api/routes/test.ts]: Thanks, I dint know about that comment.

File: api/index.ts [hunk 3]
```diff
@@ -18,8 +22,25 @@ const start = async () => {
...
+    audience: process.env.AUTH0_AUDIENCE
+  });
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+  void fastify.register(jwtAuthz);
+
+  // Auth0 plugin
+  void fastify.register(fastifyAuth0, {
+    domain: process.env.AUTH0_DOMAIN,
+    audience: process.env.AUTH0_AUDIENCE
+  });
```
Comment (ojeytonwilliams) [inline api/index.ts:37]:
```suggestion
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Cheers, Niraj, just one more thing:

**Comment by Nirajn2311** [comment]
I couldn't get the tests for the plugin code to work with `node:test` as it was not finding the required types even though I set the `typeRoots` in the `api/tsconfig.json` folder

**Comment by Nirajn2311** [comment]
@ShaunSHamilton The test command currently runs against the first test file it comes across instead of all of them.

File: api/index.ts [hunk 4]
```diff
@@ -18,8 +21,18 @@ const start = async () => {
...
+  void fastify.register(jwtAuthz);
+
   void fastify.use('/test', testMiddleware);
 
+  // Hooks
+  void fastify.addHook('preValidation', auth0Verify);
+
```
Comment (ShaunSHamilton) [inline api/index.ts:35]: I am glad this adds authentication, but I got confused why I could no longer `curl localhost:3000`:

```bash
$ curl localhost:3000
{"statusCode":401,"error":"Unauthorized","message":"Missing Authorization HTTP header."}
````

I realise I can add the header, but I would prefer to not set `auth0Verify` to be blanket on all routes. Soon, we will add `GET /ping`, and that should not be gated.

Comment (raisedadead) [reply api/index.ts:35]: This is expected. We would (like we currently do on the existing API) block all routes by default and then add exceptions to certain routes that do not need auth.

Deny all by default and then make exceptions with middleware.

Comment (ShaunSHamilton) [reply api/index.ts:35]: Sounds less fun than my way, but fine.

**Review by ojeytonwilliams — DISMISSED** [review]
Is this a separate auth flow from the one we discussed yesterday?

Apologies if I've missed the point (again), but I believe the flow was

client -> auth0 (get token, potentially including scopes) -> client -> api -> auth0 -> api -> client (encoding the session info, including scopes, as a signed session cookie)

The idea being that the client can ask the user to grant it some scopes. If the user does, that'll be communicated to the api. Finally, the api can create a session to track the fact the client is authenticated and authorized.

If that's the plan, we're missing the session stuff.

EDIT: to elaborate slightly, `fastify.authenticate` would be authenticating the _session_, rather than the token from auth0.

**Review by ShaunSHamilton — APPROVED** [review]
LGTM 👍 

_Ignoring the failed test_

File: api/index.ts [hunk 5]
```diff
@@ -18,8 +21,18 @@ const start = async () => {
...
+  void fastify.register(fastifyAuth0, {
+    domain: process.env.AUTH0_DOMAIN,
+    audience: process.env.AUTH0_AUDIENCE
+  });
+  void fastify.register(jwtAuthz);
+
   void fastify.use('/test', testMiddleware);
 
+  // Hooks
+  void fastify.addHook('preValidation', auth0Verify);
```
[the remark this answers is not in the document]
Comment (ShaunSHamilton) [reply api/index.ts:34]: See above:
https://github.com/freeCodeCamp/freeCodeCamp/pull/49378#discussion_r1115997042

File: api/index.ts (hunk 4 above)
Comment (raisedadead) [reply api/index.ts:35]: It's easier to audit endpoints that do not need "protection" than those that do. Plus, you are defaulting to zero trust, which is more secure.

Comment (ojeytonwilliams) [reply api/index.ts:35]: How do we do that with fastify, though? I haven't found a way to de-register a hook.

Comment (ojeytonwilliams) [reply api/index.ts:35]: You can achieve something similar(ish) by having two plugins "protected" and "public" and registering the hook for the "protected" one, but that's not quite the same.

File: api/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Nirajn2311) [reply api/index.ts:34]: I'm assuming that should be `fastify.authenticate` instead of `fastify.authorize`

**Comment by ojeytonwilliams** [comment]
Ooops, this shouldn't have gone in, yet, sorry about that.

Also, @raisedadead I think we can merge the next-api branch now.  Might as well simplify the workflow.

**Comment by ojeytonwilliams** [comment]
Since the context is here, this seems like the place to carry on discussing. 

The PR as written has a bit of tension with the plan we discussed most recently. To be specific, the `fastify-auth-verify` plugin (via `fastify-jwt`) does two things we care about. It decorates `fastify` with `authenticate` and, if you call `authenticate`, it decorates `req` with `user`.

The first source of tension is that we discussed using sessions to protect routes. But, in this PR, we've been talking about protecting routes via `fastify.authenticate`:

```ts
  void fastify.addHook('preValidation', auth0Verify); // where auth0Verify calls fastify.authenticate
```

If we're using Auth0 access tokens as sessions, that makes sense, but I thought we planned to handle sessions ourselves.

The second source of tension is `req.user`.  We've discussed using that like we do in the current API (i.e. have it contain the requesting user's data - challenges and so on). That's not compatible with what `fastify-auth-verify` does, i.e. putting the payload of the jwt in there.

**Comment by Nirajn2311** [comment]
> Dismissing these changes as I couldn't them to work. The request doesn't error out with a 401.

Sorry about this. I din't try async/await-ing the callback function so I dint get the intended result. Oliver's suggestion above works.

**Pull request merged**